src/common/hashes/shamap.ts

```diff
@@ -0,0 +1,263 @@
+
+/**
+ * Abstract class representing a node in a SHAMap tree.
+ *
+ * Can be either SHAMapTreeNodeInner, SHAMapTreeNodeInnerV2
+ * or SHAMapTreeNodeLeaf.
+ *
+ * @class
+ */
+export abstract class Node {
+  public constructor() {}
+  public add_item(_tag: string, _node: Node): void {
+    throw new Error('Called unimplemented virtual method SHAMapTreeNode#add_item.')
+  }
+  public get hash(): string|void {
+    throw new Error('Called unimplemented virtual method SHAMapTreeNode#hash.');
+  }
+}
+
+/**
+ * Inner (non-leaf) node in a SHAMap tree.
+ * @param {Number} depth (i.e. how many parent inner nodes)
+ * @class
+ */
+export class InnerNode extends Node {
+  public leaves: object
+  public type: 1 | 2 | 3 |4
+  public depth: number
+  public empty: boolean
+  public constructor(depth: number = 0) {
+    super()
+    this.leaves = {}
+    this.type = TYPE_INNER
+    this.depth = depth
+    this.empty = true
+  }
+  /**
+   * @param {String} tag (equates to a ledger entry `index`)
+   * @param {Node} node (to add)
+   * @return {void}
+   */  
+  public add_item(tag: string, node: Node): void {
+    const existing_node = this.get_node(tag[this.depth])
+    if (existing_node) {
+      // A node already exists in this slot
+      if (existing_node instanceof InnerNode) {
+        // There is an inner node, so we need to go deeper
+        existing_node.add_item(tag, node)
+      } else if (existing_node instanceof Leaf) {
+        if (existing_node.tag === tag) {
+          // Collision
+          throw new Error(
+              'Tried to add a node to a SHAMap that was already in there.')
+        } else {
+          // Turn it into an inner node
+          var new_inner_node = new InnerNode(this.depth + 1)
+    
+          // Parent new and existing node
+          new_inner_node.add_item(existing_node.tag, existing_node)
+          new_inner_node.add_item(tag, node)
+    
+          // And place the newly created inner node in the slot
+          this.set_node(tag[this.depth], new_inner_node)
+        }
+      }
+    } else {
+      // Neat, we have a nice open spot for the new node
+      this.set_node(tag[this.depth], node)
+    }
+  }
+  /**
+   * Overwrite the node that is currently in a given slot.
+   * @param {String} slot (a character 0-F)
+   * @param {Node} node (to place)
+   * @return {void}
+   */  
+  public set_node(slot: string, node: Node): void {
+    this.leaves[slot] = node
+    this.empty = false
+  }
+  public get_node(slot: string): Node {
+    return this.leaves[slot]
+  }
+  public get hash(): string {
+    if (this.empty) return HEX_ZERO
+    let hex = ''
+    for (let i = 0; i < 16; i++) {
+      const slot = i.toString(16).toUpperCase()
+      hex += this.leaves[slot] ? this.leaves[slot].hash : HEX_ZERO
+    }
+    const prefix = HASH_INNER_NODE.toString(16)
+    return hash(prefix + hex)
+  }  
+}
+
+/**
+ * V2 inner (non-leaf) node in a SHAMap tree.
+ * @param {Number} depth ()
```

Comment:
OOC: I've never seen this style of doc. I usually document the constructors themselves. Is there a reason you prefer it?

Reply:
This was a remnant from the ripple-hashes documentation. I didn't modify the docs.